### ISSUE TITLE
Nodes are not migrated from provision_in_progress to nodes section in Ansible Inventory when 2 or more nodes are deleted simultaneously

### ISSUE THREAD

**Issue by philipakash** [aws-quickstart/quickstart-redhat-openshift#104]
I was doing some testing with the ose_scaler and tried deleting 3/3 nodes in one go  to check the recovery. 
The Nodes  recover without any issues and even the logs show this:
(`They've additionally been removed from the provision_in_progress category`) 
but 2/3 Nodes were still under `provision_in_progress` Section .
Also Tried deleting 2/3 Nodes in one go and 1/2 were still under `provision_in_progress` Section after successful  recovery.

I know its very rare AWS will drop all three nodes but if we can handle this somehow it will be great .

**Comment by agamzn** [comment]
I'll spin up a cluster to see if I can reproduce. In the interim, can you throw the relevant logs somewhere where I can take a look?

**Comment by agamzn** [comment]
@philipakash , also a copy of `/etc/ansible/hosts` will be most helpful, too.

**Comment by philipakash** [comment]
Not sure If I can get you the logs :-( Its lost . Will post as soon as i reproduce again .
But it was pretty straight forward . Every thing showed positive except for the inventory was not true.
Let me get back with logs.

**Comment by philipakash** [comment]
@agamzn Is there an impact on the python code execution if there is a de registration delay in Nodes ?

**Comment by agamzn** [comment]
@philipakash, I think you're asking this question: "If an instance gets stuck shutting down, would that affect things?" In this context, no. The scaling code (both scale-in, and scale-out) is triggered from the cloudwatch event - in turn triggered by the ASG. By the time the python enters the stage, the autoscaling event logs already show instances scaling down.

No worries on loosing the logs. I have a theory I'm chasing down right now. If you are able to reproduce, though the inventory is really the key component I'm working against. 

As an side, I'm converting from INI to YAML in a feature branch, but that's not quite ready yet.

**Comment by philipakash** [comment]
Sounds good .
This one ? 
https://github.com/aws-quickstart/quickstart-redhat-openshift/issues/72

will this also fix the the issues which I am seeing here ?
https://github.com/aws-quickstart/quickstart-redhat-openshift/issues/107

**Comment by agamzn** [comment]
My current thinking is that #72 will address #108. We need to determine the best path forward for #107. Will discuss internally later today.  To be clear: Addressing this is #1 on my list, it's an active priority.

**Comment by philipakash** [comment]
Thank you so much @agamzn for quick reply . I will actively follow the issues.

**Comment by philipakash** [comment]
`2018-06-12T06:10:00CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:10:01CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:10:01CDT utils.py   INFO    : We're processing the scaling pipeline
2018-06-12T06:10:01CDT utils.py   INFO    : We have the following nodes to remove from the inventory:
2018-06-12T06:10:01CDT utils.py   INFO    : ['10.76.11.89']
2018-06-12T06:10:01CDT utils.py   INFO    : Removing 10.76.11.89 from category nodes
2018-06-12T06:10:01CDT utils.py   INFO    : Removing 10.76.11.89 from category new_nodes
2018-06-12T06:10:01CDT utils.py   INFO    : Removing 10.76.11.89 from category provision_in_progress
2018-06-12T06:10:01CDT utils.py   INFO    : We have the following nodes to add to the inventory:
2018-06-12T06:10:01CDT utils.py   INFO    : ['10.76.11.181']
2018-06-12T06:10:01CDT utils.py   INFO    : Complete!
2018-06-12T06:10:01CDT scaler.py  INFO    : We've detected that we need to run ansible playbooks to scale up the cluster!
2018-06-12T06:10:01CDT scaler.py  INFO    : We will run the following ansible command:
2018-06-12T06:10:01CDT scaler.py  INFO    : ansible-playbook /usr/share/ansible/openshift-ansible/scaleup_wrapper.yml --extra-vars="{'scaling_category': 'node', 'target': 'provision_in_progress'}"
2018-06-12T06:10:01CDT scaler.py  INFO    : We have 1 ansible playbooks running!
2018-06-12T06:12:01CDT scaler.py  INFO    : --------- Begin Script Invocation ---------
2018-06-12T06:12:01CDT utils.py   INFO    : Setting up the InventoryConfig Class
2018-06-12T06:12:01CDT utils.py   INFO    : Loading ansible inventory file from on-disk...
2018-06-12T06:12:01CDT utils.py   INFO    : ...Complete
2018-06-12T06:12:02CDT utils.py   INFO    : InventoryConfig setup complete!
2018-06-12T06:12:02CDT utils.py   INFO    : I'm now verifying that all required sections are present in our runtime config...
2018-06-12T06:12:02CDT utils.py   INFO    : ...Complete.
2018-06-12T06:12:02CDT utils.py   INFO    : We're populating the runtime config from data within the Ansible Inventory
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:03CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<ETCD CLSTER>>>
2018-06-12T06:12:03CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:12:03CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:12:03CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<MASTER CLSTER>>>
2018-06-12T06:12:03CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:12:03CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:12:03CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<NODES CLUSTER>>>
2018-06-12T06:12:03CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:12:04CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:12:04CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:04CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:04CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:12:04CDT scaler.py  INFO    : Verifying that the API reflects the scaling events properly
2018-06-12T06:12:04CDT scaler.py  INFO    : Great! The API contains scaling events that we need to process!
2018-06-12T06:12:04CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:12:04CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:12:04CDT utils.py   INFO    : We're processing the scaling pipeline
2018-06-12T06:12:04CDT utils.py   INFO    : We have the following nodes to remove from the inventory:
2018-06-12T06:12:04CDT utils.py   INFO    : [u'10.76.10.231']
2018-06-12T06:12:04CDT utils.py   INFO    : Removing 10.76.10.231 from category nodes
2018-06-12T06:12:04CDT utils.py   INFO    : Removing 10.76.10.231 from category new_nodes
2018-06-12T06:12:04CDT utils.py   INFO    : Removing 10.76.10.231 from category provision_in_progress
2018-06-12T06:12:04CDT utils.py   INFO    : We have the following nodes to add to the inventory:
2018-06-12T06:12:04CDT utils.py   INFO    : ['10.76.10.174']
2018-06-12T06:12:04CDT utils.py   INFO    : Complete!
2018-06-12T06:12:04CDT scaler.py  INFO    : We've detected that we need to run ansible playbooks to scale up the cluster!
2018-06-12T06:12:04CDT scaler.py  INFO    : We will run the following ansible command:
2018-06-12T06:12:04CDT scaler.py  INFO    : ansible-playbook /usr/share/ansible/openshift-ansible/scaleup_wrapper.yml --extra-vars="{'scaling_category': 'node', 'target': 'provision_in_progress'}"
2018-06-12T06:12:04CDT scaler.py  INFO    : We have 1 ansible playbooks running!
2018-06-12T06:14:10CDT scaler.py  INFO    : --------- Begin Script Invocation ---------
2018-06-12T06:14:10CDT utils.py   INFO    : Setting up the InventoryConfig Class
2018-06-12T06:14:10CDT utils.py   INFO    : Loading ansible inventory file from on-disk...
2018-06-12T06:14:10CDT utils.py   INFO    : ...Complete
2018-06-12T06:14:11CDT utils.py   INFO    : InventoryConfig setup complete!
2018-06-12T06:14:11CDT utils.py   INFO    : I'm now verifying that all required sections are present in our runtime config...
2018-06-12T06:14:11CDT utils.py   INFO    : ...Complete.
2018-06-12T06:14:11CDT utils.py   INFO    : We're populating the runtime config from data within the Ansible Inventory
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:11CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<ETCD CLSTER>>>
2018-06-12T06:14:11CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:14:12CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:14:12CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<MASTER CLSTER>>>
2018-06-12T06:14:12CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:14:12CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:14:12CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<NODES CLUSTER>>>
2018-06-12T06:14:12CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:14:13CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:14:13CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:13CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:13CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:14:13CDT scaler.py  INFO    : Verifying that the API reflects the scaling events properly
2018-06-12T06:14:13CDT scaler.py  INFO    : Great! The API contains scaling events that we need to process!
2018-06-12T06:14:13CDT utils.py   INFO    : We're processing the scaling pipeline
2018-06-12T06:14:13CDT utils.py   INFO    : No nodes were found to remove from the inventory.
2018-06-12T06:14:13CDT utils.py   INFO    : No nodes were found to add to the inventory.
2018-06-12T06:14:13CDT scaler.py  INFO    : ////////// End Script Invocation //////////
2018-06-12T06:14:30CDT scaler.py  INFO    : - A process completed. We're parsing it...
2018-06-12T06:14:30CDT utils.py   INFO    : - [nodes] playbook run results: {'failed': [], 'succeeded': [u'10.76.11.181'], 'unreachable': []}
2018-06-12T06:14:30CDT utils.py   INFO    : The json output logfile has been moved to /var/log/aws-quickstart-openshift-scaling.2018-6-12T614
2018-06-12T06:14:30CDT scaler.py  INFO    : - complete! We have 0 to go...
2018-06-12T06:14:30CDT scaler.py  INFO    : Category: nodes, Results: 1 / 0 / 0, (Succeeded / Failed / Unreachable)
2018-06-12T06:14:30CDT utils.py   INFO    : Removing 10.76.11.181 from category new_nodes
2018-06-12T06:14:30CDT utils.py   INFO    : Removing 10.76.11.181 from category provision_in_progress
2018-06-12T06:14:30CDT utils.py   INFO    : Nodes: [u'10.76.11.181'] have been permanately added to the Inventory under the nodes category
2018-06-12T06:14:30CDT utils.py   INFO    : They've additionally been removed from the provision_in_progress category
2018-06-12T06:14:30CDT scaler.py  INFO    : ////////// End Script Invocation //////////
2018-06-12T06:15:59CDT scaler.py  INFO    : --------- Begin Script Invocation ---------
2018-06-12T06:15:59CDT utils.py   INFO    : Setting up the InventoryConfig Class
2018-06-12T06:15:59CDT utils.py   INFO    : Loading ansible inventory file from on-disk...
2018-06-12T06:15:59CDT utils.py   INFO    : ...Complete
2018-06-12T06:16:00CDT utils.py   INFO    : InventoryConfig setup complete!
2018-06-12T06:16:00CDT utils.py   INFO    : I'm now verifying that all required sections are present in our runtime config...
2018-06-12T06:16:00CDT utils.py   INFO    : ...Complete.
2018-06-12T06:16:00CDT utils.py   INFO    : We're populating the runtime config from data within the Ansible Inventory
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:00CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<ETCD CLSTER>>>
2018-06-12T06:16:00CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:16:01CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:16:01CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<MASTER CLSTER>>>
2018-06-12T06:16:01CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:16:01CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:16:01CDT utils.py   INFO    : Awesome! This ASG is in the openshift <<<NODES CLUSTER>>>
2018-06-12T06:16:01CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:16:01CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:16:02CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:02CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:02CDT utils.py   INFO    : This ASG is not in the openshift cluster
2018-06-12T06:16:02CDT scaler.py  INFO    : Verifying that the API reflects the scaling events properly
2018-06-12T06:16:02CDT scaler.py  INFO    : Great! The API contains scaling events that we need to process!
2018-06-12T06:16:02CDT utils.py   INFO    : [wait_for_api]: Waiting for the EC2 API to return new instances.
2018-06-12T06:16:02CDT utils.py   INFO    : [wait_for_api] Complete
2018-06-12T06:16:02CDT utils.py   INFO    : We're processing the scaling pipeline
2018-06-12T06:16:02CDT utils.py   INFO    : We have the following nodes to remove from the inventory:
2018-06-12T06:16:02CDT utils.py   INFO    : [u'10.76.10.231']
2018-06-12T06:16:02CDT utils.py   INFO    : Removing 10.76.10.231 from category nodes
2018-06-12T06:16:02CDT utils.py   INFO    : Removing 10.76.10.231 from category new_nodes
2018-06-12T06:16:02CDT utils.py   INFO    : Removing 10.76.10.231 from category provision_in_progress
2018-06-12T06:16:02CDT utils.py   INFO    : We have the following nodes to add to the inventory:
2018-06-12T06:16:02CDT utils.py   INFO    : ['10.76.10.174']
2018-06-12T06:16:02CDT utils.py   INFO    : Complete!
2018-06-12T06:16:02CDT scaler.py  INFO    : We've detected that we need to run ansible playbooks to scale up the cluster!
2018-06-12T06:16:02CDT scaler.py  INFO    : We will run the following ansible command:
2018-06-12T06:16:02CDT scaler.py  INFO    : ansible-playbook /usr/share/ansible/openshift-ansible/scaleup_wrapper.yml --extra-vars="{'scaling_category': 'node', 'target': 'provision_in_progress'}"
2018-06-12T06:16:02CDT scaler.py  INFO    : We have 1 ansible playbooks running!
2018-06-12T06:24:47CDT scaler.py  INFO    : - A process completed. We're parsing it...
2018-06-12T06:24:48CDT utils.py   INFO    : - [nodes] playbook run results: {'failed': [], 'succeeded': [u'10.76.10.174'], 'unreachable': []}
2018-06-12T06:24:48CDT utils.py   INFO    : The json output logfile has been moved to /var/log/aws-quickstart-openshift-scaling.2018-6-12T624
2018-06-12T06:24:48CDT scaler.py  INFO    : - complete! We have 0 to go...
2018-06-12T06:24:48CDT scaler.py  INFO    : Category: nodes, Results: 1 / 0 / 0, (Succeeded / Failed / Unreachable)
2018-06-12T06:24:48CDT utils.py   INFO    : Removing 10.76.10.174 from category new_nodes
2018-06-12T06:24:48CDT utils.py   INFO    : Removing 10.76.10.174 from category provision_in_progress
2018-06-12T06:24:48CDT utils.py   INFO    : Nodes: [u'10.76.10.174'] have been permanately added to the Inventory under the nodes category
2018-06-12T06:24:48CDT utils.py   INFO    : They've additionally been removed from the provision_in_progress category
2018-06-12T06:24:48CDT scaler.py  INFO    : ////////// End Script Invocation //////////
2018-06-12T06:25:07CDT scaler.py  INFO    : - A process completed. We're parsing it...
2018-06-12T06:25:07CDT utils.py   INFO    : - [nodes] playbook run results: {'failed': [], 'succeeded': [u'10.76.10.174'], 'unreachable': []}
2018-06-12T06:25:07CDT utils.py   INFO    : The json output logfile has been moved to /var/log/aws-quickstart-openshift-scaling.2018-6-12T625
2018-06-12T06:25:07CDT scaler.py  INFO    : - complete! We have 0 to go...
2018-06-12T06:25:07CDT scaler.py  INFO    : Category: nodes, Results: 1 / 0 / 0, (Succeeded / Failed / Unreachable)
2018-06-12T06:25:07CDT utils.py   INFO    : Removing 10.76.10.174 from category new_nodes
2018-06-12T06:25:07CDT utils.py   INFO    : Removing 10.76.10.174 from category provision_in_progress
2018-06-12T06:25:07CDT utils.py   INFO    : Nodes: [u'10.76.10.174'] have been permanately added to the Inventory under the nodes category
2018-06-12T06:25:07CDT utils.py   INFO    : They've additionally been removed from the provision_in_progress category
2018-06-12T06:25:07CDT scaler.py  INFO    : ////////// End Script Invocation //////////

`

The above are the logs from latest python scaler and recorded when two nodes were deleted at the same time.
Scale up was successfull but the ansible inventory is not consistent.

`
OSEv3:
  children:
    etcd:
      hosts:
        10.76.10.205:
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.10.205
        10.76.11.189:
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.11.189
        10.76.12.190:
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.12.190
    masters:
      hosts:
        10.76.10.247: &id001
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.10.247
          openshift_node_labels:
            region: primary
            zone: default
          openshift_schedulable: 'true'
        10.76.11.251: &id002
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.11.251
          openshift_node_labels:
            region: primary
            zone: default
          openshift_schedulable: 'true'
        10.76.12.131: &id003
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.12.131
          openshift_node_labels:
            region: primary
            zone: default
          openshift_schedulable: 'true'
    new_etcd:
      hosts: {}
    new_masters:
      hosts: {}
    new_nodes:
      hosts:
        10.76.11.181: &id004
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.11.181
          openshift_node_labels:
            application_node: 'yes'
            region: infra
            registry_node: 'yes'
            router_node: 'yes'
            zone: default
    nodes:
      hosts:
        !!python/unicode '10.76.10.174':
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.10.174
          openshift_node_labels:
            application_node: 'yes'
            region: infra
            registry_node: 'yes'
            router_node: 'yes'
            zone: default
        10.76.10.247: *id001
        10.76.11.251: *id002
        10.76.12.131: *id003
        !!python/unicode '10.76.12.14':
          instance_id: i-XXXXXXXXXXX
          openshift_hostname: 10.76.12.14
          openshift_node_labels:
            application_node: 'yes'
            region: infra
            registry_node: 'yes'
            router_node: 'yes'
            zone: default
    provision_in_progress:
      hosts:
        10.76.11.181: *id004
  vars:

`
I have update the python code internally to use IPs instead of DNS.
Also `        !!python/unicode '10.76.12.14':` Is this normal?
